ciri/consensus/tip_selection/cw_rating_calculator/cw_rating_dfs_impl.c

```diff
@@ -28,55 +28,79 @@ static retcode_t cw_rating_dfs_do_dfs_from_db(tangle_t *const tangle, flex_trit_
   hash_to_indexed_hash_set_entry_t *curr_tx = NULL;
   retcode_t ret = RC_OK;
-  iota_stor_pack_t pack;
+  iota_stor_pack_t approvers_pack;
   hash243_stack_t stack = NULL;
   flex_trit_t *curr_tx_hash = NULL;
+  hash243_set_t approvees = NULL;
+  hash243_set_entry_t *iter = NULL;
+  hash243_set_entry_t *tmp = NULL;
+  bool both_approvees_exist;
+
   uint64_t start_timestamp, end_timestamp;
 
   *subtangle_size = 0;
 
   start_timestamp = current_timestamp_ms();
 
-  if ((ret = hash_pack_init(&pack, 10)) != RC_OK) {
-    goto done;
-  }
+  DECLARE_PACK_SINGLE_TX(tx, tx_ptr, transaction_pack);
 
-  if ((ret = hash243_stack_push(&stack, entry_point)) != RC_OK) {
-    goto done;
-  }
+  *subtangle_size = 0;
+
+  ERR_BIND_GOTO(hash_pack_init(&approvers_pack, 10), ret, done);
+  ERR_BIND_GOTO(hash243_stack_push(&stack, entry_point), ret, done);
 
   while (!hash243_stack_empty(stack)) {
     curr_tx_hash = hash243_stack_peek(stack);
 
     if (!hash_to_indexed_hash_set_map_contains(tx_to_approvers, curr_tx_hash)) {
-      hash_pack_reset(&pack);
-      if ((ret = iota_tangle_transaction_load_hashes_of_approvers(tangle, curr_tx_hash, &pack,
+      hash_pack_reset(&approvers_pack);
+      if ((ret = iota_tangle_transaction_load_hashes_of_approvers(tangle, curr_tx_hash, &approvers_pack,
                                                                   subtangle_before_timestamp)) != RC_OK) {
         log_error(logger_id, "Failed in loading approvers, error code is: %" PRIu64 "\n", ret);
         goto done;
       }
-      if ((ret = hash_to_indexed_hash_set_map_add_new_set(tx_to_approvers, curr_tx_hash, &curr_tx,
-                                                          (*subtangle_size)++)) != RC_OK) {
-        goto done;
-      }
+
+      ERR_BIND_GOTO(
+          hash_to_indexed_hash_set_map_add_new_set(tx_to_approvers, curr_tx_hash, &curr_tx, (*subtangle_size)++), ret,
+          done);
       hash243_stack_pop(&stack);
-      while (pack.num_loaded > 0) {
-        curr_tx_hash = ((flex_trit_t *)pack.models[--pack.num_loaded]);
-        // Add each found approver to the currently traversed tx
-        if ((ret = hash243_stack_push(&stack, curr_tx_hash)) != RC_OK) {
-          goto done;
+      while (approvers_pack.num_loaded > 0) {
+        curr_tx_hash = ((flex_trit_t *)approvers_pack.models[--approvers_pack.num_loaded]);
+
+        ERR_BIND_GOTO(iota_tangle_transaction_load_partial(tangle, curr_tx_hash, &transaction_pack,
+                                                           PARTIAL_TX_MODEL_ESSENCE_ATTACHMENT_METADATA),
+                      ret, done)
+
+        ERR_BIND_GOTO(hash243_stack_push(&stack, curr_tx_hash), ret, done);
+        both_approvees_exist = true;
+        if (!transaction_solid(&tx)) {
+          hash243_set_add(&approvees, transaction_branch(&tx));
+          hash243_set_add(&approvees, transaction_trunk(&tx));
+
+          HASH_SET_ITER(approvees, iter, tmp) {
```

Comment:
Do we really need to have the overhead of a set for only two approvees ? Can't we just do two `exist` ?